From c2c3a075ea10a0907d2ad6ce20d258a26bbe46ed Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Thu, 26 Jun 2025 13:11:53 +0200
Subject: [PATCH 083/115] Input: novatek-nvt-ts - convert to regmap API

In preparation for SPI support, convert the Novatek NVT touchscreen
driver to use the regmap API.

Signed-off-by: Dinolek <git@dinolek.me>
---
 drivers/input/touchscreen/Kconfig             |  16 +-
 drivers/input/touchscreen/Makefile            |   3 +-
 ...novatek-nvt-ts.c => novatek-nvt-ts-core.c} | 156 ++++--------------
 .../input/touchscreen/novatek-nvt-ts-i2c.c    |  66 ++++++++
 drivers/input/touchscreen/novatek-nvt-ts.h    |  54 ++++++
 5 files changed, 164 insertions(+), 131 deletions(-)
 rename drivers/input/touchscreen/{novatek-nvt-ts.c => novatek-nvt-ts-core.c} (59%)
 create mode 100644 drivers/input/touchscreen/novatek-nvt-ts-i2c.c
 create mode 100644 drivers/input/touchscreen/novatek-nvt-ts.h

diff --git a/drivers/input/touchscreen/Kconfig b/drivers/input/touchscreen/Kconfig
index 9b9ae8ac3..8b11f87d7 100644
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -691,15 +691,23 @@ config TOUCHSCREEN_MTOUCH
 	  To compile this driver as a module, choose M here: the
 	  module will be called mtouch.
 
-config TOUCHSCREEN_NOVATEK_NVT_TS
-	tristate "Novatek NT11205 touchscreen support"
+
+config TOUCHSCREEN_NOVATEK_NVT_TS_CORE
+	tristate
+
+config TOUCHSCREEN_NOVATEK_NVT_TS_I2C
+	tristate "Novatek I2C touchscreen support"
 	depends on I2C
+	select REGMAP_I2C
+	select TOUCHSCREEN_NOVATEK_NVT_TS_CORE
 	help
-	  Say Y here if you have a Novatek NT11205 touchscreen.
+	  Say Y here if you have a Novatek touchscreen connected to
+	  your system via I2C.
+
 	  If unsure, say N.
 
 	  To compile this driver as a module, choose M here: the
-	  module will be called novatek-nvt-ts.
+	  module will be called novatek-nvt-ts-i2c.
 
 config TOUCHSCREEN_IMAGIS
 	tristate "Imagis touchscreen support"
diff --git a/drivers/input/touchscreen/Makefile b/drivers/input/touchscreen/Makefile
index bfd9de833..25d7d736a 100644
--- a/drivers/input/touchscreen/Makefile
+++ b/drivers/input/touchscreen/Makefile
@@ -68,7 +68,8 @@ obj-$(CONFIG_TOUCHSCREEN_MIGOR)		+= migor_ts.o
 obj-$(CONFIG_TOUCHSCREEN_MMS114)	+= mms114.o
 obj-$(CONFIG_TOUCHSCREEN_MSG2638)	+= msg2638.o
 obj-$(CONFIG_TOUCHSCREEN_MTOUCH)	+= mtouch.o
-obj-$(CONFIG_TOUCHSCREEN_NOVATEK_NVT_TS)	+= novatek-nvt-ts.o
+obj-$(CONFIG_TOUCHSCREEN_NOVATEK_NVT_TS_CORE)	+= novatek-nvt-ts-core.o
+obj-$(CONFIG_TOUCHSCREEN_NOVATEK_NVT_TS_I2C)	+= novatek-nvt-ts-i2c.o
 obj-$(CONFIG_TOUCHSCREEN_HP600)		+= hp680_ts_input.o
 obj-$(CONFIG_TOUCHSCREEN_HP7XX)		+= jornada720_ts.o
 obj-$(CONFIG_TOUCHSCREEN_IPAQ_MICRO)	+= ipaq-micro-ts.o
diff --git a/drivers/input/touchscreen/novatek-nvt-ts.c b/drivers/input/touchscreen/novatek-nvt-ts-core.c
similarity index 59%
rename from drivers/input/touchscreen/novatek-nvt-ts.c
rename to drivers/input/touchscreen/novatek-nvt-ts-core.c
index 0f771f681..63a7f88e3 100644
--- a/drivers/input/touchscreen/novatek-nvt-ts.c
+++ b/drivers/input/touchscreen/novatek-nvt-ts-core.c
@@ -1,44 +1,19 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 /*
- * Driver for Novatek NT11205 i2c touchscreen controller as found
- * on the Acer Iconia One 7 B1-750 tablet.
+ * Driver for Novatek touchscreen controller.
  *
  * Copyright (c) 2023 Hans de Goede <hdegoede@redhat.com>
  */
 
 #include <linux/delay.h>
 #include <linux/gpio/consumer.h>
-#include <linux/interrupt.h>
-#include <linux/i2c.h>
-#include <linux/input.h>
 #include <linux/input/mt.h>
-#include <linux/input/touchscreen.h>
+#include <linux/interrupt.h>
 #include <linux/module.h>
-
+#include <linux/property.h>
 #include <linux/unaligned.h>
 
-#define NVT_TS_TOUCH_START		0x00
-#define NVT_TS_TOUCH_SIZE		6
-
-#define NVT_TS_PARAMETERS_START		0x78
-/* These are offsets from NVT_TS_PARAMETERS_START */
-#define NVT_TS_PARAMS_WIDTH		0x04
-#define NVT_TS_PARAMS_HEIGHT		0x06
-#define NVT_TS_PARAMS_MAX_TOUCH		0x09
-#define NVT_TS_PARAMS_MAX_BUTTONS	0x0a
-#define NVT_TS_PARAMS_IRQ_TYPE		0x0b
-#define NVT_TS_PARAMS_CHIP_ID		0x0e
-#define NVT_TS_PARAMS_SIZE		0x0f
-
-#define NVT_TS_MAX_TOUCHES		10
-#define NVT_TS_MAX_SIZE			4096
-
-#define NVT_TS_TOUCH_INVALID		0xff
-#define NVT_TS_TOUCH_SLOT_SHIFT		3
-#define NVT_TS_TOUCH_TYPE_MASK		GENMASK(2, 0)
-#define NVT_TS_TOUCH_NEW		1
-#define NVT_TS_TOUCH_UPDATE		2
-#define NVT_TS_TOUCH_RELEASE		3
+#include "novatek-nvt-ts.h"
 
 static const int nvt_ts_irq_type[4] = {
 	IRQF_TRIGGER_RISING,
@@ -47,56 +22,16 @@ static const int nvt_ts_irq_type[4] = {
 	IRQF_TRIGGER_HIGH
 };
 
-struct nvt_ts_i2c_chip_data {
-	u8 chip_id;
-};
-
-struct nvt_ts_data {
-	struct i2c_client *client;
-	struct input_dev *input;
-	struct gpio_desc *reset_gpio;
-	struct regulator_bulk_data regulators[2];
-	struct touchscreen_properties prop;
-	int max_touches;
-	u8 buf[NVT_TS_TOUCH_SIZE * NVT_TS_MAX_TOUCHES];
-};
-
-static int nvt_ts_read_data(struct i2c_client *client, u8 reg, u8 *data, int count)
-{
-	struct i2c_msg msg[2] = {
-		{
-			.addr = client->addr,
-			.len = 1,
-			.buf = &reg,
-		},
-		{
-			.addr = client->addr,
-			.flags = I2C_M_RD,
-			.len = count,
-			.buf = data,
-		}
-	};
-	int ret;
-
-	ret = i2c_transfer(client->adapter, msg, ARRAY_SIZE(msg));
-	if (ret != ARRAY_SIZE(msg)) {
-		dev_err(&client->dev, "Error reading from 0x%02x: %d\n", reg, ret);
-		return (ret < 0) ? ret : -EIO;
-	}
-
-	return 0;
-}
-
 static irqreturn_t nvt_ts_irq(int irq, void *dev_id)
 {
 	struct nvt_ts_data *data = dev_id;
-	struct device *dev = &data->client->dev;
+	const struct device *dev = data->dev;
 	int i, error, slot, x, y;
 	bool active;
 	u8 *touch;
 
-	error = nvt_ts_read_data(data->client, NVT_TS_TOUCH_START, data->buf,
-				 data->max_touches * NVT_TS_TOUCH_SIZE);
+	error = regmap_raw_read(data->regmap, NVT_TS_TOUCH_START, data->buf,
+				data->max_touches * NVT_TS_TOUCH_SIZE);
 	if (error)
 		return IRQ_HANDLED;
 
@@ -147,11 +82,11 @@ static int nvt_ts_start(struct input_dev *dev)
 
 	error = regulator_bulk_enable(ARRAY_SIZE(data->regulators), data->regulators);
 	if (error) {
-		dev_err(&data->client->dev, "failed to enable regulators\n");
+		dev_err(data->dev, "failed to enable regulators\n");
 		return error;
 	}
 
-	enable_irq(data->client->irq);
+	enable_irq(data->irq);
 	gpiod_set_value_cansleep(data->reset_gpio, 0);
 
 	return 0;
@@ -161,14 +96,14 @@ static void nvt_ts_stop(struct input_dev *dev)
 {
 	struct nvt_ts_data *data = input_get_drvdata(dev);
 
-	disable_irq(data->client->irq);
+	disable_irq(data->irq);
 	gpiod_set_value_cansleep(data->reset_gpio, 1);
 	regulator_bulk_disable(ARRAY_SIZE(data->regulators), data->regulators);
 }
 
 static int nvt_ts_suspend(struct device *dev)
 {
-	struct nvt_ts_data *data = i2c_get_clientdata(to_i2c_client(dev));
+	const struct nvt_ts_data *data = dev_get_drvdata(dev);
 
 	guard(mutex)(&data->input->mutex);
 
@@ -180,7 +115,7 @@ static int nvt_ts_suspend(struct device *dev)
 
 static int nvt_ts_resume(struct device *dev)
 {
-	struct nvt_ts_data *data = i2c_get_clientdata(to_i2c_client(dev));
+	const struct nvt_ts_data *data = dev_get_drvdata(dev);
 
 	guard(mutex)(&data->input->mutex);
 
@@ -190,17 +125,17 @@ static int nvt_ts_resume(struct device *dev)
 	return 0;
 }
 
-static DEFINE_SIMPLE_DEV_PM_OPS(nvt_ts_pm_ops, nvt_ts_suspend, nvt_ts_resume);
+EXPORT_GPL_SIMPLE_DEV_PM_OPS(nvt_ts_pm_ops, nvt_ts_suspend, nvt_ts_resume);
 
-static int nvt_ts_probe(struct i2c_client *client)
+int nvt_ts_probe(struct device *dev, int irq, struct regmap *regmap,
+		 const struct input_id *id)
 {
-	struct device *dev = &client->dev;
 	int error, width, height, irq_type;
 	struct nvt_ts_data *data;
-	const struct nvt_ts_i2c_chip_data *chip;
+	const struct nvt_ts_chip_data *chip;
 	struct input_dev *input;
 
-	if (!client->irq) {
+	if (!irq) {
 		dev_err(dev, "Error no irq specified\n");
 		return -EINVAL;
 	}
@@ -209,12 +144,14 @@ static int nvt_ts_probe(struct i2c_client *client)
 	if (!data)
 		return -ENOMEM;
 
-	chip = device_get_match_data(&client->dev);
+	chip = device_get_match_data(dev);
 	if (!chip)
 		return -EINVAL;
 
-	data->client = client;
-	i2c_set_clientdata(client, data);
+	data->dev = dev;
+	data->regmap = regmap;
+	data->irq = irq;
+	dev_set_drvdata(dev, data);
 
 	/*
 	 * VCC is the analog voltage supply
@@ -244,8 +181,8 @@ static int nvt_ts_probe(struct i2c_client *client)
 
 	/* Wait for controller to come out of reset before params read */
 	msleep(100);
-	error = nvt_ts_read_data(data->client, NVT_TS_PARAMETERS_START,
-				 data->buf, NVT_TS_PARAMS_SIZE);
+	error = regmap_raw_read(data->regmap, NVT_TS_PARAMETERS_START,
+				data->buf, NVT_TS_PARAMS_SIZE);
 	gpiod_set_value_cansleep(data->reset_gpio, 1); /* Put back in reset */
 	regulator_bulk_disable(ARRAY_SIZE(data->regulators), data->regulators);
 	if (error)
@@ -275,8 +212,8 @@ static int nvt_ts_probe(struct i2c_client *client)
 	if (!input)
 		return -ENOMEM;
 
-	input->name = client->name;
-	input->id.bustype = BUS_I2C;
+	input->name = dev_name(dev);
+	input->id = *id;
 	input->open = nvt_ts_start;
 	input->close = nvt_ts_stop;
 
@@ -292,10 +229,10 @@ static int nvt_ts_probe(struct i2c_client *client)
 	data->input = input;
 	input_set_drvdata(input, data);
 
-	error = devm_request_threaded_irq(dev, client->irq, NULL, nvt_ts_irq,
+	error = devm_request_threaded_irq(dev, irq, NULL, nvt_ts_irq,
 					  IRQF_ONESHOT | IRQF_NO_AUTOEN |
 						nvt_ts_irq_type[irq_type],
-					  client->name, data);
+					  dev_name(dev), data);
 	if (error) {
 		dev_err(dev, "failed to request irq: %d\n", error);
 		return error;
@@ -309,41 +246,8 @@ static int nvt_ts_probe(struct i2c_client *client)
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(nvt_ts_probe);
 
-static const struct nvt_ts_i2c_chip_data nvt_nt11205_ts_data = {
-	.chip_id = 0x05,
-};
-
-static const struct nvt_ts_i2c_chip_data nvt_nt36672a_ts_data = {
-	.chip_id = 0x08,
-};
-
-static const struct of_device_id nvt_ts_of_match[] = {
-	{ .compatible = "novatek,nt11205-ts", .data = &nvt_nt11205_ts_data },
-	{ .compatible = "novatek,nt36672a-ts", .data = &nvt_nt36672a_ts_data },
-	{ }
-};
-MODULE_DEVICE_TABLE(of, nvt_ts_of_match);
-
-static const struct i2c_device_id nvt_ts_i2c_id[] = {
-	{ .name = "nt11205-ts", .driver_data = (unsigned long)&nvt_nt11205_ts_data },
-	{ .name = "nt36672a-ts", .driver_data = (unsigned long)&nvt_nt36672a_ts_data },
-	{ }
-};
-MODULE_DEVICE_TABLE(i2c, nvt_ts_i2c_id);
-
-static struct i2c_driver nvt_ts_driver = {
-	.driver = {
-		.name	= "novatek-nvt-ts",
-		.pm	= pm_sleep_ptr(&nvt_ts_pm_ops),
-		.of_match_table = nvt_ts_of_match,
-	},
-	.probe = nvt_ts_probe,
-	.id_table = nvt_ts_i2c_id,
-};
-
-module_i2c_driver(nvt_ts_driver);
-
-MODULE_DESCRIPTION("Novatek NT11205 touchscreen driver");
+MODULE_DESCRIPTION("Novatek core touchscreen driver");
 MODULE_AUTHOR("Hans de Goede <hdegoede@redhat.com>");
 MODULE_LICENSE("GPL");
diff --git a/drivers/input/touchscreen/novatek-nvt-ts-i2c.c b/drivers/input/touchscreen/novatek-nvt-ts-i2c.c
new file mode 100644
index 000000000..ec6ed3468
--- /dev/null
+++ b/drivers/input/touchscreen/novatek-nvt-ts-i2c.c
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Driver for Novatek touchscreen controller.
+ *
+ * Copyright (c) 2023 Hans de Goede <hdegoede@redhat.com>
+ */
+
+#include <linux/i2c.h>
+
+#include "novatek-nvt-ts.h"
+
+static const struct regmap_config nvt_ts_i2c_regmap_conf = {
+	.reg_bits = 8,
+	.val_bits = 8,
+};
+
+static const struct input_id nvt_ts_i2c_input_id = {
+	.bustype = BUS_I2C,
+};
+
+static int nvt_ts_i2c_probe(struct i2c_client *i2c)
+{
+	struct regmap *regmap =
+		devm_regmap_init_i2c(i2c, &nvt_ts_i2c_regmap_conf);
+	if (IS_ERR(regmap))
+		return PTR_ERR(regmap);
+
+	return nvt_ts_probe(&i2c->dev, i2c->irq, regmap, &nvt_ts_i2c_input_id);
+}
+
+static const struct nvt_ts_chip_data nvt_nt11205_ts_data = {
+	.chip_id = 0x05,
+};
+
+static const struct nvt_ts_chip_data nvt_nt36672a_ts_data = {
+	.chip_id = 0x08,
+};
+
+static const struct of_device_id nvt_ts_i2c_of_match[] = {
+	{ .compatible = "novatek,nt11205-ts", .data = &nvt_nt11205_ts_data },
+	{ .compatible = "novatek,nt36672a-ts", .data = &nvt_nt36672a_ts_data },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, nvt_ts_i2c_of_match);
+
+static const struct i2c_device_id nvt_ts_i2c_id[] = {
+	{ "nt11205-ts", (unsigned long)&nvt_nt11205_ts_data },
+	{ "nt36672a-ts", (unsigned long)&nvt_nt36672a_ts_data },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(i2c, nvt_ts_i2c_id);
+
+static struct i2c_driver nvt_ts_i2c_driver = {
+	.driver = {
+		.name	= "novatek-nvt-ts-i2c",
+		.pm	= pm_sleep_ptr(&nvt_ts_pm_ops),
+		.of_match_table = nvt_ts_i2c_of_match,
+	},
+	.probe = nvt_ts_i2c_probe,
+	.id_table = nvt_ts_i2c_id,
+};
+module_i2c_driver(nvt_ts_i2c_driver);
+
+MODULE_DESCRIPTION("Novatek i2c touchscreen driver");
+MODULE_AUTHOR("Hans de Goede <hdegoede@redhat.com>");
+MODULE_LICENSE("GPL");
diff --git a/drivers/input/touchscreen/novatek-nvt-ts.h b/drivers/input/touchscreen/novatek-nvt-ts.h
new file mode 100644
index 000000000..bde2f2158
--- /dev/null
+++ b/drivers/input/touchscreen/novatek-nvt-ts.h
@@ -0,0 +1,54 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+#ifndef __NOVATEK_NVT_TS_H_
+#define __NOVATEK_NVT_TS_H_
+
+#include <linux/input.h>
+#include <linux/input/touchscreen.h>
+#include <linux/regmap.h>
+#include <linux/regulator/consumer.h>
+
+#define NVT_TS_TOUCH_START		0x00
+#define NVT_TS_TOUCH_SIZE		6
+
+#define NVT_TS_PARAMETERS_START		0x78
+/* These are offsets from NVT_TS_PARAMETERS_START */
+#define NVT_TS_PARAMS_WIDTH		0x04
+#define NVT_TS_PARAMS_HEIGHT		0x06
+#define NVT_TS_PARAMS_MAX_TOUCH		0x09
+#define NVT_TS_PARAMS_MAX_BUTTONS	0x0a
+#define NVT_TS_PARAMS_IRQ_TYPE		0x0b
+#define NVT_TS_PARAMS_CHIP_ID		0x0e
+#define NVT_TS_PARAMS_SIZE		0x0f
+
+#define NVT_TS_MAX_TOUCHES		10
+#define NVT_TS_MAX_SIZE			4096
+
+#define NVT_TS_TOUCH_INVALID		0xff
+#define NVT_TS_TOUCH_SLOT_SHIFT		3
+#define NVT_TS_TOUCH_TYPE_MASK		GENMASK(2, 0)
+#define NVT_TS_TOUCH_NEW		1
+#define NVT_TS_TOUCH_UPDATE		2
+#define NVT_TS_TOUCH_RELEASE		3
+
+struct nvt_ts_data {
+	struct device *dev;
+	struct regmap *regmap;
+	struct input_dev *input;
+	struct gpio_desc *reset_gpio;
+	struct regulator_bulk_data regulators[2];
+	struct touchscreen_properties prop;
+	int max_touches;
+	int irq;
+	u8 buf[NVT_TS_TOUCH_SIZE * NVT_TS_MAX_TOUCHES];
+};
+
+struct nvt_ts_chip_data {
+	u8 chip_id;
+};
+
+int nvt_ts_probe(struct device *dev, int irq, struct regmap *regmap,
+		 const struct input_id *id);
+
+extern const struct dev_pm_ops nvt_ts_pm_ops;
+
+#endif
-- 
2.56.0

